pkg/storage/azure.go

```diff
@@ -0,0 +1,446 @@
+	if err != nil {
+		return nil, err
+	}
+
+	var signingKeys *core.SigningKeys
+	if pt == internalProviderType {
+		signingKeys, err = s.SigningKeys(ctx, provider.Namespace)
+	} else if pt == mirrorProviderType {
+		signingKeys, err = s.MirroredSigningKeys(ctx, provider.Hostname, provider.Namespace)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	provider.Filename = path.Base(archivePath)
+	provider.SigningKeys = *signingKeys
+	return provider, nil
+}
+
+func (s *AzureStorage) GetProvider(ctx context.Context, namespace, name, version, os, arch string) (*core.Provider, error) {
+	p, err := s.getProvider(ctx, internalProviderType, &core.Provider{
+		Namespace: namespace,
+		Name:      name,
+		Version:   version,
+		OS:        os,
+		Arch:      arch,
+	})
+	return p, err
+}
+
+func (s *AzureStorage) GetMirroredProvider(ctx context.Context, provider *core.Provider) (*core.Provider, error) {
+	return s.getProvider(ctx, mirrorProviderType, provider)
+}
+
+func (s *AzureStorage) listProviderVersions(ctx context.Context, pt providerType, provider *core.Provider) ([]*core.Provider, error) {
+	prefix := providerStoragePrefix(s.prefix, pt, provider.Hostname, provider.Namespace, provider.Name)
+
+	var providers []*core.Provider
+	pager := s.client.NewListBlobsFlatPager(s.container, &azblob.ListBlobsFlatOptions{
+		Prefix: &prefix,
+	})
+	for pager.More() {
+		page, err := pager.NextPage(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("failed to page next page: %w", err)
+		}
+
+		for _, obj := range page.Segment.BlobItems {
+			p, err := core.NewProviderFromArchive(filepath.Base(*obj.Name))
+			if err != nil {
+				continue
+			}
+
+			if provider.Version != "" && provider.Version != p.Version {
+				continue
+			}
+
+			p.Hostname = provider.Hostname
+			p.Namespace = provider.Namespace
+			archiveUrl, err := s.presignedURL(ctx, *obj.Name)
+			if err != nil {
+				return nil, err
+			}
+			p.DownloadURL = archiveUrl
+
+			providers = append(providers, &p)
+		}
+	}
+
+	return providers, nil
+}
+
+func (s *AzureStorage) ListProviderVersions(ctx context.Context, namespace, name string) (*core.ProviderVersions, error) {
+	providers, err := s.listProviderVersions(ctx, internalProviderType, &core.Provider{Namespace: namespace, Name: name})
+	if err != nil {
+		return nil, err
+	}
+
+	collection := NewCollection()
+	for _, p := range providers {
+		collection.Add(p)
+	}
+	return collection.List(), nil
+}
+
+func (s *AzureStorage) ListMirroredProviders(ctx context.Context, provider *core.Provider) ([]*core.Provider, error) {
+	return s.listProviderVersions(ctx, mirrorProviderType, provider)
+}
+
+func (s *AzureStorage) UploadProviderReleaseFiles(ctx context.Context, namespace, name, filename string, file io.Reader) error {
+	if namespace == "" {
+		return fmt.Errorf("namespace argument is empty")
+	}
+
+	if name == "" {
+		return fmt.Errorf("name argument is empty")
+	}
+
+	if filename == "" {
+		return fmt.Errorf("name argument is empty")
```

Comment:
Should be `filename argument is empty`, but it's also incorrect in `s3.go` and `gcs.go`

Reply:
Resolved in 4b2455ef1811e096e987d0446b3c1e2981614246